README.md
# SimpleFreeForum


CHANGELOG.md
# Revision history for forum

## 0.1.0.0 -- YYYY-mm-dd

* First version. Released on an unsuspecting world.


src/DB.hs
{-# LANGUAGE EmptyDataDecls             #-}
{-# LANGUAGE FlexibleContexts           #-}
{-# LANGUAGE GADTs                      #-}
{-# LANGUAGE GeneralizedNewtypeDeriving #-}
{-# LANGUAGE MultiParamTypeClasses      #-}
{-# LANGUAGE OverloadedStrings          #-}
{-# LANGUAGE QuasiQuotes                #-}
{-# LANGUAGE TemplateHaskell            #-}
{-# LANGUAGE TypeFamilies               #-}
{-# LANGUAGE DerivingStrategies #-}
{-# LANGUAGE StandaloneDeriving #-}
{-# LANGUAGE UndecidableInstances #-}
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE FlexibleInstances #-}
{-# OPTIONS_GHC -Wno-deprecations #-}
{-# LANGUAGE TypeApplications #-}
{-# LANGUAGE RankNTypes #-}


module DB where

import Types
import           Control.Monad.IO.Class                     (liftIO)
import           Control.Monad.Logger                       (runStderrLoggingT, MonadLogger, runFileLoggingT)
import           Control.Monad.Reader
import           Data.Text
import           Data.Time
import           Database.Esqueleto.Experimental
import           Database.Persist.Postgresql                (ConnectionString, withPostgresqlConn, withPostgresqlPool, PostgresConf (PostgresConf))
import           Database.Persist.TH                        ( AtLeastOneUniqueKey(..)
       , OnlyOneUniqueKey(..)
       , mkMigrate
       , mkPersist
       , persistLowerCase
       , share
       , sqlSettings
       )

-- TYPES

share [mkPersist sqlSettings, mkMigrate "migrateAll"] [persistLowerCase|
DbComment 
    topic Text 
    text Text
    time UTCTime
    deriving Show
|]


toComment :: DbComment -> Comment 
toComment (DbComment topic text time) = Comment (Topic topic) (CommentText text) time


fromComment :: Comment -> DbComment
fromComment (Comment (Topic topic) (CommentText text) time) = DbComment topic text time

-- DB

queryGetComments :: (MonadIO m, MonadLogger m)
         => Text -> SqlReadT m [Entity DbComment]
queryGetComments t =
    select $ do
    c <- from $ table @DbComment
    where_ (c ^. DbCommentTopic ==. val t)
    pure c


queryGetTopics :: (MonadIO m, MonadLogger m) => SqlReadT m [Value Text]
queryGetTopics =
    select $ distinct $ do
        c <- from $ table @DbComment
        let topics = c ^. DbCommentTopic
        return (c ^. DbCommentTopic)


conn = "host=localhost dbname=forum_db user=tester password=test_password port=5432"

{-
data DBEnv = DBEnv {
    logfile :: String,
    connection :: String
}

newtype AppDB a = AppDB { runAppDB' :: ReaderT DBEnv IO a }

runAppDB :: DBEnv -> AppDB a -> IO a
runAppDB env appDB = runReaderT (runAppDB' appDB) env 

-}

runQuery logfile conn query = runFileLoggingT logfile $ withPostgresqlConn conn (runSqlConn query)


setupDb :: (MonadIO m, MonadLogger m)
          => SqlPersistT m ()
setupDb = do
  runMigration migrateAll

initDB :: IO ()
initDB = runQuery "log.txt" conn setupDb

getTopics :: IO [Topic]
getTopics = do
    a <- liftIO $ runQuery "log.txt" conn queryGetTopics
    let c = unValue <$> a
    return $ Topic <$> c


getComment :: Topic -> IO [Comment]
getComment t = do
    a <- runFileLoggingT "log.txt" $ withPostgresqlConn conn (runSqlConn $ queryGetComments (getTopic t))
    let c = entityVal <$> a
    return $ toComment <$> c


addComment :: Comment -> IO ()
addComment c = do
    runFileLoggingT "log.txt" $ withPostgresqlConn conn (runSqlConn $ insert (fromComment c))
    return ()

src/Core.hs
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE TypeOperators #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE GeneralizedNewtypeDeriving #-}
{-# LANGUAGE InstanceSigs #-}

module Core where

import Servant.API
import Servant.Server
import Servant
import Network.Wai.Handler.Warp
import Control.Monad.Except
import Control.Monad.Reader

import Text.Blaze.Html (stringComment)
import qualified Data.ByteString.Lazy as BL
import Data.Aeson
import Data.Time
import Data.Text as T
import DB
import Types


instance ToHttpApiData Topic where
  toUrlPiece = getTopic 

instance FromHttpApiData Topic where
  parseUrlPiece = Right . Topic



type GetTopicListAPI   = "list" :> Get '[JSON] [Topic]
type GetCommentListAPI = Capture "topic" Topic :> "view" :> Get '[JSON] [Comment]
type PostCommentAPI    = Capture "topic" Topic :> "add"  :> ReqBody '[JSON] CommentText :> Post '[JSON] Topic

type ForumAPI = GetTopicListAPI
           :<|> GetCommentListAPI
           :<|> PostCommentAPI


forumAPI :: Proxy ForumAPI
forumAPI = Proxy


-- newtype Handler a = IO (Either ServerError a)
-- runHandler' :: ExceptT ServerError IO a

-- Reader String a -> Handler a

-- type Server api = ServerT api Handler

forumServer :: ServerT ForumAPI (AppM ServerError)
forumServer = handleGetTopicList
         :<|> handleGetCommentList
         :<|> handlePostComment

handleGetTopicList :: (AppM ServerError) [Topic]
handleGetTopicList = liftIO getTopics

handleGetCommentList :: Topic -> (AppM ServerError) [Comment]
handleGetCommentList t = do
    com <- liftIO $ getComment t
    return com
    where
        custom404Err = err404 { errBody = "myfile.txt just isn't there, please leave this server alone." }
--    return [Comment (CommentId 4) (Topic "ciao") (CommentText "hello") t]

handlePostComment :: Topic -> CommentText -> (AppM ServerError) Topic
handlePostComment topic comText = do
  time <- liftIO getCurrentTime 
  let c = Comment topic comText time
  liftIO $ addComment c
  return topic


funToHandler :: (AppM ServerError) a -> Handler a
funToHandler (AppM t) = Handler $ runReaderT t "env"

main = run 8080 $ serve forumAPI $ hoistServer forumAPI funToHandler forumServer


---
data Error = Error
type Env = String

newtype AppM e a = AppM (ReaderT Env (ExceptT e IO) a)
  deriving (Functor, Applicative, Monad, MonadError e, MonadReader Env, MonadIO)


runAppM :: AppM e a -> Env -> IO (Either e a)
runAppM (AppM m) env = runExceptT $ runReaderT m env



src/Core1.hs

{-# LANGUAGE DataKinds #-}
{-# LANGUAGE TypeOperators #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE GeneralizedNewtypeDeriving #-}

module Core1 where

import Servant.API
import Servant.Server
import Servant
import Network.Wai.Handler.Warp
import Control.Monad.Except
import Control.Monad.Reader

import Types
import Text.Blaze.Html (stringComment)
import qualified Data.ByteString.Lazy as BL
import Data.Aeson
import Data.Time
import Data.Text as T

type GetTopicListAPI   = "list" :> Get '[JSON] [Topic]
type GetCommentListAPI = "view" :> ReqBody '[JSON] Topic :> Post '[JSON] [Comment]
type PostCommentAPI    = "add"  :> ReqBody '[JSON] Comment :> Post '[JSON] Topic

type ForumAPI = GetTopicListAPI
           :<|> GetCommentListAPI
           :<|> PostCommentAPI


forumAPI :: Proxy ForumAPI
forumAPI = Proxy



-- newtype Handler a = IO (Either ServerError a)
-- runHandler' :: ExceptT ServerError IO a

-- Reader String a -> Handler a

-- type Server api = ServerT api Handler

--handleGetTopicList :: Handler [Topic]
handleGetTopicList :: Handler [Topic]
handleGetTopicList = return [Topic "ciao"]

handleGetCommentList :: Topic -> Handler [Comment]
handleGetCommentList t = do
    stringcomm <- liftIO (BL.readFile "comments.txt")
    let mcomm = decode stringcomm
    case mcomm of
        Nothing -> throwError custom404Err
        Just c  -> return [c]

    where
        custom404Err = err404 { errBody = "myfile.txt just isn't there, please leave this server alone." }

--handlePostComment :: Comment -> Handler Topic
handlePostComment c = return $ Topic "ai"


forumServer :: Server ForumAPI
forumServer = handleGetTopicList
         :<|> handleGetCommentList
         :<|> handlePostComment

main = run 8080 (serve forumAPI forumServer)

--- test

forumServer' :: ServerT ForumAPI (ReaderT String (ExceptT ServerError IO))
forumServer' = handleGetTopicList'
         :<|> handleGetCommentList'
         :<|> handlePostComment'

handleGetTopicList' :: ReaderT String (ExceptT ServerError IO) [Topic]
handleGetTopicList' = return [Topic "cioa"]

handleGetCommentList' :: Topic -> ReaderT String (ExceptT ServerError IO) [Comment]
handleGetCommentList' t = do
    stringcomm <- liftIO (BL.readFile "comments.txt")
    t <- liftIO getCurrentTime
    return [Comment (Topic "ciao") (CommentText "hello") t]

handlePostComment' :: Comment -> ReaderT String (ExceptT ServerError IO) Topic
handlePostComment' c = asks (Topic . T.pack)

funToHandler :: (ReaderT String (ExceptT ServerError IO)) a -> Handler a
funToHandler t = Handler $ runReaderT t "env" 

main' = run 8080 $ serve forumAPI $ hoistServer forumAPI funToHandler forumServer'


---
data Error = Error
data Env = Env

newtype AppM e a = AppM (ReaderT Env (ExceptT e IO) a)
  deriving (Functor, Applicative, Monad, MonadError e, MonadReader Env)


runAppM :: AppM e a -> Env -> IO (Either e a)
runAppM (AppM m) env = runExceptT $ runReaderT m env


{-
conn = "host=localhost dbname=forum_db user=tester password=test_password port=5432"

aliftSqlPersistMPool y x = liftIO (runSqlPersistMPool y x)


mmain :: IO [Comment]
--mmain = runStderrLoggingT $ withPostgresqlPool conn 10 $ aliftSqlPersistMPool $ do
mmain = runStderrLoggingT $ withPostgresqlPool conn 10 $ \b -> liftIO (runSqlPersistMPool (do
        runMigration migrateAll
        now <- liftIO getCurrentTime
        personId <- insert $ Comment "test" "questo e' un altro commento" now
        comments <- (entityVal <$>) <$> queryGetComments "lol"
        topics <-  (unValue <$>) <$> queryGetTopics
        liftIO $ print topics
        liftIO (print comments)
        return comments) b)
-}